Moved and revised the value types article (#16765)
dotnet/docs: docs/csharp/language-reference/builtin-types/value-types.md

---
title: "Value types - C# reference"
ms.date: 01/22/2020
f1_keywords: 
  - "cs.valuetypes"
helpviewer_keywords: 
  - "value types [C#]"
  - "types [C#], value types"
  - "C# language, value types"
ms.assetid: 471eb994-2958-49d5-a6be-19b4313f80a3
---
# Value types (C# reference)

*Value types* and [reference types](../keywords/reference-types.md) are the two main categories of C# types. A variable of a value type contains an instance of the type. This differs from a variable of a reference type, which contains a reference to an instance of the type. By default, on [assignment](../operators/assignment-operator.md), passing an argument to a method, or returning a method result, variable values are copied. In the case of value-type variables, the corresponding type instances are copied. The following example demonstrates that behavior:

> [!code-csharp[copy of values](~/samples/csharp/language-reference/builtin-types/ValueTypes.cs#ValueTypeCopied)]

Reviewer: Should the new code reference syntax be used ?

https://docs.microsoft.com/en-us/contribute/how-to-write-use-markdown#snippet-syntax-reference

```suggestion
:::code language="csharp" source="~/samples/csharp/language-reference/builtin-types/ValueTypes.cs" id="ValueTypeCopied":::
```

> As the preceding example shows, operations on a value-type variable don't affect other variables.

Reviewer 2: I'd like to rephrase this so it highlights the distinction between *variables* and *the thing a variable refers to*:

```suggestion
As the preceding example shows, operations on a value-type variable affect only that instance of the value type, stored in the variable.
```

If a value type contains a data member of a reference type, only the reference to the instance of the reference type is copied when a value-type instance is copied. Both the copy and original value-type instance have access to the same reference-type instance. The following example demonstrates that behavior:

[!code-csharp[shallow copy](~/samples/csharp/language-reference/builtin-types/ValueTypes.cs#ShallowCopy)]

> [!NOTE]
> To make your code less error-prone and more robust, define and use immutable value types. This article uses mutable value types only for demonstration purposes.

## Kinds of value types

A value type can be one of the two following kinds:

- a [structure type](../keywords/struct.md), which encapsulates data and related functionality
- an [enumeration type](enum.md), which is defined by a set of named constants and represents a choice or a combination of choices

A [nullable value type](nullable-value-types.md) `T?` represents all values of its underlying value type `T` and an additional [null](../keywords/null.md) value. You cannot assign `null` to a variable of a value type, unless it's a nullable value type.

> ## Built-in value types. Simple types

Reviewer: Looks better to me with parenthesis.

```suggestion
## Built-in value types (simple types)
```

Reviewer: Got you. Then, what do you think about having simple types as h3 under the Built-in value types ?
That way, it looks more clearer

Author: That would work. But now I think if the docs even should introduce and describe simple types. The link to spec would work if necessary. @BillWagner what do you think?

Reviewer 2: I think the heading should be just "Built in". This section covers both, and the distinction isn't needed for the heading:

```suggestion
## Built-in value types
```
